src/lint/validators.nim

```diff
@@ -176,10 +176,38 @@ func isUuidV4*(s: string): bool =
 
 var seenUuids = initHashSet[string](250)
 
+proc extractPlaceholder(value: string): string =
+  var i = 0
+  var phStart = -1
+  var ph = ""
+  while i < value.len:
+    var c = value[i]
+    if phStart == -1:
+      if c == '%' and value[i+1] == '{':
+        phStart = i+2
+        inc i
+    else:
+      if c == '}':
+        return ph
+      else:
+        ph = ph & c
+    inc i
+  return ""
+
+func isFilesPattern*(s: string): bool =
+  if not isEmptyOrWhitespace(s):
+    let ph = extractPlaceholder(s)
```

Comment:
Theoretically speaking, people could be using multiple placeholders in the files pattern. It is not likely, but it is something to consider supporting.

Reply:
Well, there could possibly be a path like: `%{pascal_slug}/%{snake_slug}.vb` or whatever. I've just checked all the track's `.files` properties, and there _is_ one track (Swift) that uses: `"Tests/%{pascal_slug}Tests/%{pascal_slug}Tests.swift"`

So we do have to support multiple placeholders.

> Also what about error cases? Presumably it's an error to add a placeholder which doesn't exist, so %{foobar} is an error

That is correct. That would be an error.

> [...] but how about incomplete placeholders? Is %{foobar an error, or valid?

Maybe it can be a warning?

Reply:
Added a low-tech warning in https://github.com/exercism/configlet/pull/435/commits/60e26ada55557742ea5d275221e7bef191b11813.

I might refactor later - e.g. `isFilesPattern` could be renamed and return an enum like `fpOk`, `fpIsBlank`, `fpHasInvalidPlaceholder`, `fpContainsForbiddenCharacter`.